src/Lemon/Containers/Converters/Ncch2Binary.cs

```diff
@@ -0,0 +1,237 @@
+
+            writer.Stream.Position += 0x10; // Reserved
+
+            WriteSHA256(writer, root.Children["logo.bin"]);
+            writer.Write(source.Header.ProductCode.PadRight(0x10, '\0'));
+            WriteSHA256(writer, root.Children["extended_header"]);
+            writer.Write((int)(root.Children["extended_header"]?.Stream?.Length ?? 0));
+
+            writer.Stream.Position += 0x4; // Reserved
+
+            for (int i = 0; i < source.Header.Flags.Length; i++) {
+                byte flag = source.Header.Flags[i];
+                writer.Write(flag);
+            }
+
+            // Now we begin writing the NCCH content
+            writer.Stream.Position = 0x200;
+            WriteFile(writer, root.Children["extended_header"], false);
+            WriteFile(writer, root.Children["access_descriptor"], false);
+
+            writer.Stream.Position = 0x190;
+            WriteOffsetSizeAndData(writer, root.Children["sdk_info.txt"], source.Header.SdkInfoOffset);
+            WriteOffsetSizeAndData(writer, root.Children["logo.bin"], source.Header.LogoOffset);
+            WriteOffsetSizeAndData(writer, root.Children["system"], source.Header.SystemOffset, source.Header.SystemHashSize, true);
+
+            writer.Stream.Position += 0x4; // Reserved
+            WriteOffsetSizeAndData(writer, root.Children["rom"], source.Header.RomOffset, source.Header.RomHashSize, true);
+
+            writer.Stream.Position += 0x4; // Reserved
+
+            WriteSHA256(writer, root.Children["system"], source.Header.SystemHashSize, true);
+            WriteSHA256(writer, root.Children["rom"], source.Header.RomHashSize, true);
+
+            // Write the full size of the binary (in units) in 0x104
+            writer.Stream.Position = 0x104;
+            writer.Write((int)writer.Stream.Length / NcchHeader.Unit);
+
+            return binary;
+        }
+
+        void WriteFile(DataWriter writer, Node file, bool isRequired = true)
+        {
+            if (file == null && isRequired) {
+                throw new System.IO.FileNotFoundException("Missing NCCH file");
+            } else if (file == null) {
+                return;
+            }
+
+            if (file.Format is not IBinary) {
+                throw new FormatException("Cannot write file as it is not binary");
+            }
+
+            file.Stream.WriteTo(writer.Stream);
+        }
+
+        void WriteOffsetSizeAndData(DataWriter writer, Node file, int offsetInUnits, int hashRegion = 0, bool hasHashRegion = false)
+        {
+            if (file != null) {
+                long position = writer.Stream.Position;
+                int offset = offsetInUnits * NcchHeader.Unit;
+                int fileLength = (int)file.Stream.Length;
+                int fileLengthInUnits = fileLength / NcchHeader.Unit;
+
+                bool usedChunk = false;
+                var fileRange = Enumerable.Range(offset, fileLength);
+                foreach (var usedUnit in usedUnits) {
```

Comment:
What's the goal of this dictionary?

Reply:
This dictionary contains the starting offset and the length of already written chunks, so later it can be used to determine before writing a new chunk if there was content in that range